File: ifc-tools/src/oebf/ifc_importer.py
import json
import re
import uuid
from pathlib import Path
from datetime import date

import ifcopenshell
import ifcopenshell.geom


IFC_TO_OEBF = {
    "IfcWall": "IfcWall",
    "IfcWallStandardCase": "IfcWall",
    "IfcSlab": "IfcSlab",
    "IfcBeam": "IfcBeam",
    "IfcColumn": "IfcColumn",
    "IfcRoof": "IfcRoof",
}


def import_ifc(ifc_path: Path, out_dir: Path) -> None:
    model = ifcopenshell.open(str(ifc_path))
    out_dir.mkdir(parents=True, exist_ok=True)

    for sub in ["paths", "profiles", "elements", "materials", "junctions", "arrays", "symbols", "groups", "schema", "ifc"]:
        (out_dir / sub).mkdir(exist_ok=True)

    projects = model.by_type("IfcProject")
    project = projects[0] if projects else None
    project_name = project.Name if project and project.Name else ifc_path.stem

    elements = []
    for ifc_type, oebf_type in IFC_TO_OEBF.items():
        for entity in model.by_type(ifc_type):
            element_id = _slugify(entity.GlobalId or str(uuid.uuid4()))
            result = _process_element(entity, element_id, oebf_type, out_dir)
            if result:
                elements.append(element_id)

    _write_manifest(out_dir, project_name)
    _write_model(out_dir, elements)
    _write_materials(out_dir, model)


def _process_element(entity, element_id, oebf_type, out_dir):
    """Write path + element JSON for one IFC entity. Returns element_id on success."""
    try:
        # Try to extract swept geometry; fall back to a unit-length placeholder path.
        path_id = f"path-{element_id}"
        segment = _extract_path_segment(entity)
        path_data = {
            "$schema": "oebf://schema/0.1/path",
            "id": path_id,
            "type": "Path",
            "description": f"Imported path for {entity.is_a()} {getattr(entity, 'Name', '') or ''}".strip(),
            "closed": False,
            "segments": [segment],
            "tags": ["imported"],
        }
        (out_dir / "paths" / f"{path_id}.json").write_text(json.dumps(path_data, indent=2))

        elem_data = {
            "$schema": "oebf://schema/0.1/element",
            "id": element_id,
            "type": "Element",
            "description": getattr(entity, "Name", None) or entity.is_a(),
            "ifc_type": oebf_type,
            "path_id": path_id,
            "profile_id": "profile-imported-placeholder",
            "sweep_mode": "perpendicular",
            "cap_start": "flat",
            "cap_end": "flat",
            "start_offset": 0.0,
            "end_offset": 0.0,
            "properties": {"imported_from_ifc": True},
        }
        (out_dir / "elements" / f"{element_id}.json").write_text(json.dumps(elem_data, indent=2))
        return element_id
    except Exception as exc:
        print(f"  Warning: could not process {entity.is_a()} {entity.GlobalId}: {exc}")
        return None


def _extract_path_segment(entity):
    """Return a line segment dict for the entity's swept axis, or a 1 m placeholder."""
    try:
        rep = entity.Representation
        if rep:
            for item in rep.Representations:
                for shape_item in item.Items:
                    # IfcExtrudedAreaSolid gives us axis + direction
                    if shape_item.is_a("IfcExtrudedAreaSolid"):
                        pos = shape_item.Position
                        if pos and pos.Location:
                            loc = pos.Location.Coordinates
                            x, y = float(loc[0]), float(loc[1])
                            z = float(loc[2]) if len(loc) > 2 else 0.0
                            depth = float(shape_item.Depth)
                            # Extrusion direction in local coords; apply to start point
                            d = shape_item.ExtrudedDirection.DirectionRatios
                            dx, dy, dz = float(d[0]) * depth, float(d[1]) * depth, float(d[2]) * depth
                            return {
                                "type": "line",
                                "start": {"x": round(x, 4), "y": round(y, 4), "z": round(z, 4)},
                                "end": {"x": round(x + dx, 4), "y": round(y + dy, 4), "z": round(z + dz, 4)},
                            }
    except Exception:
        pass

    # Fallback: 1 m stub along X at origin
    return {
        "type": "line",
        "start": {"x": 0.0, "y": 0.0, "z": 0.0},
        "end": {"x": 1.0, "y": 0.0, "z": 0.0},
    }


def _write_manifest(out_dir: Path, project_name: str) -> None:
    manifest = {
        "format": "oebf",
        "format_version": "0.1.0",
        "project_name": project_name,
        "description": "Imported from IFC",
        "created": str(date.today()),
        "units": "metres",
        "coordinate_system": "right_hand_z_up",
        "files": {
            "model": "model.json",
            "materials": "materials/library.json",
            "schema": "schema/oebf-schema.json",
        },
    }
    (out_dir / "manifest.json").write_text(json.dumps(manifest, indent=2))


def _write_model(out_dir: Path, elements: list) -> None:
    model_data = {
        "hierarchy": {
            "type": "Project",
            "id": "project-root",
            "description": "Imported",
            "children": [],
        },
        "elements": elements,
        "objects": [],
        "arrays": [],
        "junctions": [],
    }
    (out_dir / "model.json").write_text(json.dumps(model_data, indent=2))


def _write_materials(out_dir: Path, ifc_model) -> None:
    ifc_materials = ifc_model.by_type("IfcMaterial")
    materials = []
    seen: set[str] = set()
    for m in ifc_materials:
        if m.Name in seen:
            continue
        seen.add(m.Name)
        materials.append({
            "id": f"mat-{_slugify(m.Name)}",
            "type": "Material",
            "name": m.Name,
            "category": "imported",
            "colour_hex": "#888888",
            "ifc_material_name": m.Name,
            "properties": {},
            "interactions": {},
        })
    (out_dir / "materials" / "library.json").write_text(json.dumps({"materials": materials}, indent=2))


def _slugify(text: str) -> str:
    text = text.lower().strip()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    return text.strip("-")[:40]

File: viewer/src/ifc/ifcImporter.js
/**
 * ifcImporter.js — Convert IFC STEP text to OEBF entities.
 *
 * Writes path, element, and slab JSON files to the bundle via adapter.
 * Returns { projectName, elementIds, slabIds, materials }.
 */

import { parseStep } from './stepParser.js';

const IFC_TO_OEBF_TYPE = {
  IFCWALL:              'IfcWall',
  IFCWALLSTANDARDCASE: 'IfcWall',
  IFCSLAB:             'IfcSlab',
  IFCROOF:             'IfcRoof',
  IFCBEAM:             'IfcBeam',
  IFCCOLUMN:           'IfcColumn',
};

const SLAB_TYPES = new Set(['IFCSLAB', 'IFCROOF']);

/**
 * Import an IFC file (as text) into the current bundle via adapter.
 *
 * @param {string} text  — IFC STEP file content
 * @param {object} adapter  — FsaAdapter or MemoryAdapter
 * @returns {Promise<{projectName:string, elementIds:string[], slabIds:string[], materials:object[]}>}
 */
export async function importIfcText(text, adapter) {
  const entities = parseStep(text);

  // Project name
  let projectName = 'Imported Project';
  for (const [, e] of entities) {
    if (e.type === 'IFCPROJECT' && typeof e.attrs[2] === 'string') {
      projectName = e.attrs[2];
      break;
    }
  }

  const elementIds = [];
  const slabIds    = [];
  const materials  = [];
  const seenMatIds = new Set();

  // Collect IfcMaterial entries
  for (const [, e] of entities) {
    if (e.type === 'IFCMATERIAL') {
      const name = typeof e.attrs[0] === 'string' ? e.attrs[0] : null;
      if (name) {
        const id = `mat-${_slugify(name)}`;
        if (!seenMatIds.has(id)) {
          seenMatIds.add(id);
          materials.push({
            id,
            type: 'Material',
            name,
            category: 'imported',
            colour_hex: '#888888',
            ifc_material_name: name,
            properties: {},
            interactions: {},
          });
        }
      }
    }
  }

  // Process elements
  for (const [, e] of entities) {
    const oebfType = IFC_TO_OEBF_TYPE[e.type];
    if (!oebfType) continue;

    const globalId = typeof e.attrs[0] === 'string' ? e.attrs[0] : _makeId();
    const name     = typeof e.attrs[2] === 'string' ? e.attrs[2] : e.type;

    // IFC Representation is at attr index 6 for IfcElement subtypes
    const repRef = e.attrs[6] || e.attrs[5];

    const baseId = `imp-${_slugify(globalId).slice(0, 22)}`;
    const elemId = baseId;
    const pathId = `path-${baseId}`;

    const segment = _extractPathSegment(entities, repRef);

    await adapter.writeJson(`paths/${pathId}.json`, {
      $schema:     'oebf://schema/0.1/path',
      id:          pathId,
      type:        'Path',
      description: `Imported path for ${name}`,
      closed:      false,
      segments:    [segment],
      tags:        ['imported'],
    });

    if (SLAB_TYPES.has(e.type)) {
      await adapter.writeJson(`slabs/${elemId}.json`, {
        $schema:          'oebf://schema/0.1/slab',
        id:               elemId,
        type:             'Slab',
        description:      name,
        ifc_type:         oebfType,
        boundary_path_id: pathId,
        thickness_m:      0.2,
        material_id:      '',
        parent_group_id:  '',
        properties:       { imported_from_ifc: true },
      });
      slabIds.push(elemId);
    } else {
      await adapter.writeJson(`elements/${elemId}.json`, {
        $schema:        'oebf://schema/0.1/element',
        id:             elemId,
        type:           'Element',
        description:    name,
        ifc_type:       oebfType,
        path_id:        pathId,
        profile_id:     'profile-imported-placeholder',
        sweep_mode:     'perpendicular',
        cap_start:      'flat',
        cap_end:        'flat',
        start_offset:   0.0,
        end_offset:     0.0,
        parent_group_id: '',
        properties:     { imported_from_ifc: true },
      });
      elementIds.push(elemId);
    }
  }

  return { projectName, elementIds, slabIds, materials };
}

/**
 * Traverse the IFC entity graph from a Representation ref to extract a path segment.
 * Falls back to a 1 m stub along X if geometry cannot be extracted.
 */
function _extractPathSegment(entities, repRef) {
  const fallback = {
    type:  'line',
    start: { x: 0, y: 0, z: 0 },
    end:   { x: 1, y: 0, z: 0 },
  };

  try {
    if (!repRef || typeof repRef !== 'object' || repRef.type !== 'ref') return fallback;

    // repRef → IfcProductDefinitionShape
    const prodShape = entities.get(repRef.id);
    if (!prodShape) return fallback;

    // IfcProductDefinitionShape: Name(0), Description(1), Representations(2)
    const reps = prodShape.attrs[2];
    if (!Array.isArray(reps)) return fallback;

    for (const repItemRef of reps) {
      if (!repItemRef || repItemRef.type !== 'ref') continue;
      const shapeRep = entities.get(repItemRef.id); // IfcShapeRepresentation
      if (!shapeRep) continue;

      // IfcShapeRepresentation: ContextOfItems(0), RepresentationIdentifier(1),
      //                         RepresentationType(2), Items(3)
      const items = shapeRep.attrs[3];
      if (!Array.isArray(items)) continue;

      for (const itemRef of items) {
        if (!itemRef || itemRef.type !== 'ref') continue;
        const solid = entities.get(itemRef.id);
        if (!solid || solid.type !== 'IFCEXTRUDEDAREASOLID') continue;

        // IfcExtrudedAreaSolid: SweptArea(0), Position(1), ExtrudedDirection(2), Depth(3)
        const depth = typeof solid.attrs[3] === 'number' ? solid.attrs[3] : 1.0;

        // Start position from Position (IfcAxis2Placement3D)
        let sx = 0, sy = 0, sz = 0;
        const posRef = solid.attrs[1];
        if (posRef && posRef.type === 'ref') {
          const placement = entities.get(posRef.id);
          if (placement) {
            const locRef = placement.attrs[0];
            if (locRef && locRef.type === 'ref') {
              const pt = entities.get(locRef.id);
              if (pt) {
                const coords = pt.attrs[0];
                if (Array.isArray(coords)) {
                  sx = typeof coords[0] === 'number' ? coords[0] : 0;
                  sy = typeof coords[1] === 'number' ? coords[1] : 0;
                  sz = typeof coords[2] === 'number' ? coords[2] : 0;
                }
              }
            }
          }
        }

        // Direction from ExtrudedDirection (IfcDirection)
        let dx = 1, dy = 0, dz = 0;
        const dirRef = solid.attrs[2];
        if (dirRef && dirRef.type === 'ref') {
          const dir = entities.get(dirRef.id);
          if (dir) {
            const ratios = dir.attrs[0];
            if (Array.isArray(ratios)) {
              dx = typeof ratios[0] === 'number' ? ratios[0] : 1;
              dy = typeof ratios[1] === 'number' ? ratios[1] : 0;
              dz = typeof ratios[2] === 'number' ? ratios[2] : 0;
            }
          }
        }

        return {
          type:  'line',
          start: { x: _r4(sx),           y: _r4(sy),           z: _r4(sz) },
          end:   { x: _r4(sx + dx*depth), y: _r4(sy + dy*depth), z: _r4(sz + dz*depth) },
        };
      }
    }
  } catch { /* fall through to fallback */ }

  return fallback;
}

function _r4(n)       { return Math.round(n * 10000) / 10000; }
function _makeId()    { return Math.random().toString(36).slice(2, 10); }
function _slugify(s)  { return String(s).toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '').slice(0, 40); }
